### PR TITLE
Fix windows sam parsing

### PR DESCRIPTION
**Pull request by JSCU-CNI**
This fix depends on fox-it/dissect.regf#9.

### CODE REVIEW

**Comment by Schamper** [comment]
Also fixes https://github.com/fox-it/dissect/issues/17

File: dissect/target/helpers/regutil.py [hunk 1]
```diff
@@ -72,6 +72,11 @@ def name(self) -> str:
         """Returns the name of this key."""
         raise NotImplementedError()
 
+    @property
+    def class_name(self) -> str:
+        """Returns the classname of this key."""
```
Comment (Schamper) [inline dissect/target/helpers/regutil.py:77]:
```suggestion
        """Returns the class name of this key."""
```

File: dissect/target/helpers/regutil.py [hunk 2]
```diff
@@ -276,6 +282,9 @@ def add_subkey(self, name: str, key: str):
         """Add a subkey to this key."""
         self._subkeys[name.lower()] = key
 
+    def add_classname(self, value: str):
+        self._class_name = value
+
```
Comment (Schamper) [inline dissect/target/helpers/regutil.py:287]:
```suggestion
```
Perhaps better to do this as just a constructor argument to `VirtualKey`. See the other suggestion.

File: dissect/target/helpers/regutil.py [hunk 3]
```diff
@@ -264,6 +269,7 @@ class VirtualKey(RegistryKey):
     def __init__(self, hive: RegistryHive, path: str):
         self._path: str = path
         self._name: str = path.split("\\")[-1]
+        self._class_name: str = None
```
Comment (Schamper) [inline dissect/target/helpers/regutil.py:272]:
```suggestion
    def __init__(self, hive: RegistryHive, path: str, class_name: Optional[str] = None):
        self._path: str = path
        self._name: str = path.split("\\")[-1]
        self._class_name: str = class_name
```

File: dissect/target/plugins/os/windows/sam.py [hunk 4]
```diff
@@ -137,83 +210,225 @@
...
+        aqwerty = b"!@#$%^&*()qwertyUIOPAzxcvbnmQQQQQQQQQQQQ)(*@&%\0"
+        anum = b"0123456789012345678901234567890123456789\0"
+
+        f_reg = self.target.registry.key(self.SAM_KEY).value("F").value
+        f = c_sam.DOMAIN_ACCOUNT_F(f_reg)
+        f_key = f_reg[len(c_sam.DOMAIN_ACCOUNT_F) :]
+        fk = c_sam.SAM_KEY(f_key)
+
+        if f.revision not in [0x02, 0x03]:
+            raise PluginError(f"Unsupported Domain Account F revision encountered: {f.revision}")
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:269]:
```suggestion
            raise ValueError(f"Unsupported Domain Account F revision encountered: {f.revision}")
```

File: dissect/target/plugins/os/windows/sam.py [hunk 5]
```diff
@@ -137,83 +210,225 @@
...
+
+        if fk.revision not in [0x01, 0x02]:
+            raise PluginError(f"Unsupported SAM Key Data revision encountered: {fk.revision}")
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:272]:
```suggestion
            raise ValueError(f"Unsupported SAM Key Data revision encountered: {fk.revision}")
```

File: dissect/target/plugins/os/windows/sam.py [hunk 6]
```diff
@@ -137,83 +210,225 @@
...
+
+        if fk.revision == 0x01:  # SAM key revision 0x01 involving RC4 (samsrv.dll: KEDecryptKeyWithRC4)
+            rc4_key = MD5.new(fk.salt + aqwerty + syskey + anum).digest()
+            samkey_data = ARC4.new(rc4_key).encrypt(fk.key + fk.checksum)
+            samkey = samkey_data[:16]
+            checksum = samkey_data[16:]
+
+            if checksum != MD5.new(samkey + anum + samkey + aqwerty).digest():
+                raise PluginError("SAM key checksum validation failed!")
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:281]:
```suggestion
                raise ValueError("SAM key checksum validation failed!")
```

File: dissect/target/plugins/os/windows/sam.py [hunk 7]
```diff
@@ -1,28 +1,32 @@
+from struct import pack
+from typing import Iterator, Tuple
+
+from Crypto.Cipher import AES, ARC4, DES
+from Crypto.Hash import MD5, SHA256
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:5]: Can you use `hashlib` for the MD5 and SHA256 calculations?

File: dissect/target/plugins/os/windows/sam.py [hunk 8]
```diff
@@ -137,83 +210,225 @@
...
+            fk = c_sam.SAM_KEY_AES(f_key)
+            key_data = f_key[len(c_sam.SAM_KEY_AES) : len(c_sam.SAM_KEY_AES) + fk.data_len]
+            checksum_data = f_key[
+                len(c_sam.SAM_KEY_AES) + fk.data_len : len(c_sam.SAM_KEY_AES) + fk.data_len + fk.checksum_len
+            ]
+            samkey = SamPlugin.perform_aes(key=syskey, value=key_data, iv=fk.salt)[:16]
+            checksum = SamPlugin.perform_aes(key=syskey, value=checksum_data, iv=fk.salt)[:32]
+
+            if checksum != SHA256.new(samkey).digest():
+                raise PluginError("SAM key checksum validation failed!")
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:294]:
```suggestion
                raise ValueError("SAM key checksum validation failed!")
```

File: dissect/target/plugins/os/windows/sam.py [hunk 9]
```diff
@@ -137,83 +210,225 @@
...
+            return samkey
+
+    @staticmethod
+    def expand_des_key(key: bytes) -> bytes:
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:298]: Could you move this to the global scope and out of the class?

File: dissect/target/plugins/os/windows/sam.py [hunk 10]
```diff
@@ -137,83 +210,225 @@
...
+                ((key[3] & 0x0F) << 3 | ((key[4] >> 5) & 0x07)) << 1,
+                ((key[4] & 0x1F) << 2 | ((key[5] >> 6) & 0x03)) << 1,
+                ((key[5] & 0x3F) << 1 | ((key[6] >> 7) & 0x01)) << 1,
+                (key[6] & 0x7F) << 1,
+            ]
+        )
+        return bytes(s)
+
+    @staticmethod
+    def rid_to_key(rid: int) -> Tuple[bytes, bytes]:
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:316]: Could you move this to the global scope and out of the class?

Comment (Schamper) [reply dissect/target/plugins/os/windows/sam.py:316]:
```suggestion
    def rid_to_key(rid: int) -> tuple[bytes, bytes]:
```
Also small suggestion to go with that :)

File: dissect/target/plugins/os/windows/sam.py [hunk 11]
```diff
@@ -137,83 +210,225 @@
...
+        s = rid.to_bytes(4, "little", signed=False)
+        k1 = SamPlugin.expand_des_key(bytes([s[0], s[1], s[2], s[3], s[0], s[1], s[2]]))
+        k2 = SamPlugin.expand_des_key(bytes([s[3], s[0], s[1], s[2], s[3], s[0], s[1]]))
+
+        return k1, k2
+
+    @staticmethod
+    def perform_aes(key: bytes, value: bytes, iv: bytes) -> bytes:
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:324]: Could you move this to the global scope and out of the class? And maybe rename to `aes_crypt` or `aes_decrypt`.

File: dissect/target/plugins/os/windows/sam.py [hunk 12]
```diff
@@ -137,83 +210,225 @@
...
+        plaintext = b""
+        cipher = AES.new(key, AES.MODE_CBC, IV=iv)
+        n = 16
+        for block in [value[i : i + n] for i in range(0, len(value), n)]:
+            plaintext += cipher.decrypt(block)
+
+        return plaintext
+
+    @staticmethod
+    def decrypt_single_hash(rid: int, samkey: bytes, enc_hash: bytes, apwd: bytes) -> bytes:
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:334]: Could you move this to the global scope and out of the class?

File: dissect/target/plugins/os/windows/sam.py [hunk 13]
```diff
@@ -137,83 +210,225 @@
...
+        sh = c_sam.SAM_HASH(enc_hash)
+
+        if sh.revision not in [0x01, 0x02]:
+            raise PluginError(f"Unsupported LM/NT hash revision encountered: {sh.revision}")
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:338]:
```suggestion
            raise ValueError(f"Unsupported LM/NT hash revision encountered: {sh.revision}")
```

File: dissect/target/plugins/os/windows/sam.py [hunk 14]
```diff
@@ -137,83 +210,225 @@
...
         """
-        for users_key in self.target.registry.iterkeys(self.KEY):
+
+        syskey = self.retrieve_syskey()  # aka. bootkey
+        samkey = self.calculate_samkey(syskey)  # aka. hashed bootkey or hbootkey
+
+        almpassword = b"LMPASSWORD\0"
+        antpassword = b"NTPASSWORD\0"
+
+        for users_key in self.target.registry.iterkeys(f"{self.SAM_KEY}\\Users"):
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:393]:
```suggestion
        for users_key in self.target.registry.keys(f"{self.SAM_KEY}\\Users"):
```

**Review by Schamper — CHANGES_REQUESTED** [review]

**Comment by Schamper** [comment]
Can you change the dependency on `dissect.regf` in `setup.py` to be:
```
"dissect.regf>=3.3.dev,<4.0.dev",
```

File: dissect/target/plugins/os/windows/sam.py [hunk 15]
```diff
@@ -137,83 +210,225 @@
...
+
+
+def aes_decrypt(key: bytes, value: bytes, iv: bytes) -> bytes:
+    plaintext = b""
+    cipher = AES.new(key, AES.MODE_CBC, IV=iv)
+    n = 16
+    for block in [value[i : i + n] for i in range(0, len(value), n)]:
+        plaintext += cipher.decrypt(block)
+
+    return plaintext
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/sam.py:261]: Isn't this the same as:
```python
return AES.new(key, AES.MODE_CBC, iv).decrypt(value)
```

I believe pycryptodome does the block loop automatically.

File: tests/test_plugins_os_windows_sam.py [hunk 16]
```diff
@@ -0,0 +1,348 @@
+from Crypto.Hash import MD4
+from flow.record.fieldtypes import datetime as dt
+
+from dissect.target.helpers.regutil import VirtualKey
+from dissect.target.plugins.os.windows.sam import SamPlugin
```
Comment (Schamper) [inline tests/test_plugins_os_windows_sam.py:5]:
```suggestion
from flow.record.fieldtypes import datetime as dt

from dissect.target.helpers.regutil import VirtualKey

try:
    from Crypto.Hash import MD4
    from dissect.target.plugins.os.windows.sam import SamPlugin
    
    HAVE_CRYPTO = True
except ImportError:
    HAVE_CRYPTO = False
```

File: tests/test_plugins_os_windows_sam.py [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+    assert results[2].flags == 20
+    assert results[2].countrycode == 1
+    assert results[2].failedlogins == 1
+    assert results[2].logins == 8
+    assert results[2].lm == "b0e5df88ef1c16f1dde347abf49e1d97"
+    assert results[2].nt == MD4.new("L0ngLiv3LM!".encode("utf-16-le")).digest().hex()
+
+
+def test_sam_plugin_rev2(target_win_users, hive_hklm):
+
```
Comment (Schamper) [inline tests/test_plugins_os_windows_sam.py:139]:
```suggestion
@pytest.mark.skipif(not HAVE_CRYPTO, reason="requires pycryptodome")
def test_sam_plugin_rev2(target_win_users, hive_hklm):
```

**Review by Schamper — CHANGES_REQUESTED** [review]

**Comment by Schamper** [comment]
I tried to push some final changes myself but it didn't let me, I guess the checkbox to do so wasn't ticked for this PR?

Could you apply these final changes? It's to add `pycryptodome` to the `[full]` install of `dissect.target`, fixing the tests and a small style change in the test:

```diff
diff --git a/setup.py b/setup.py
index 275d82f..f488694 100644
--- a/setup.py
+++ b/setup.py
@@ -79,6 +79,7 @@ setup(
             "dissect.xfs>=3.0.dev,<4.0.dev",
             "ipython",
             "fusepy",
+            "pycryptodome",
             "pyyaml",
             "yara-python",
             # dissect.target's caching uses flow.record functionlity which depends on the
diff --git a/tests/test_plugins_os_windows_sam.py b/tests/test_plugins_os_windows_sam.py
index aa15bc8..bb952b6 100644
--- a/tests/test_plugins_os_windows_sam.py
+++ b/tests/test_plugins_os_windows_sam.py
@@ -1,22 +1,24 @@
+import pytest
 from flow.record.fieldtypes import datetime as dt
 
 from dissect.target.helpers.regutil import VirtualKey
 
 try:
     from Crypto.Hash import MD4
+
     from dissect.target.plugins.os.windows.sam import SamPlugin
-    
+
     HAVE_CRYPTO = True
 except ImportError:
     HAVE_CRYPTO = False
 
-sam_key_name = "SAM\\SAM\\Domains\\Account"
-system_key_name = "SYSTEM\\ControlSet001\\Control\\LSA"
+SAM_KEY_PATH = "SAM\\SAM\\Domains\\Account"
+SYSTEM_KEY_PATH = "SYSTEM\\ControlSet001\\Control\\LSA"
 
 
 @pytest.mark.skipif(not HAVE_CRYPTO, reason="requires pycryptodome")
 def test_sam_plugin_rev1(target_win_users, hive_hklm):
-    sam_key = VirtualKey(hive_hklm, sam_key_name)
+    sam_key = VirtualKey(hive_hklm, SAM_KEY_PATH)
     sam_key.add_value(
         "F",
         bytes.fromhex(
@@ -107,10 +109,10 @@ def test_sam_plugin_rev1(target_win_users, hive_hklm):
     users_key.add_subkey("000003E8", e8_key)
 
     sam_key.add_subkey("Users", users_key)
-    hive_hklm.map_key(sam_key_name, sam_key)
+    hive_hklm.map_key(SAM_KEY_PATH, sam_key)
 
     # Create SYSTEM keys
-    system_key = VirtualKey(hive_hklm, system_key_name)
+    system_key = VirtualKey(hive_hklm, SYSTEM_KEY_PATH)
     data_key = VirtualKey(hive_hklm, "Data", class_name="6d1ae431")
     system_key.add_subkey("Data", data_key)
     gbg_key = VirtualKey(hive_hklm, "GBG", class_name="1ddbd1f5")
@@ -119,7 +121,7 @@ def test_sam_plugin_rev1(target_win_users, hive_hklm):
     system_key.add_subkey("JD", jd_key)
     skew1_key = VirtualKey(hive_hklm, "Skew1", class_name="c42b803c")
     system_key.add_subkey("Skew1", skew1_key)
-    hive_hklm.map_key(system_key_name, system_key)
+    hive_hklm.map_key(SYSTEM_KEY_PATH, system_key)
 
     target_win_users.add_plugin(SamPlugin)
     results = list(target_win_users.sam())
@@ -143,7 +145,7 @@ def test_sam_plugin_rev1(target_win_users, hive_hklm):
 
 @pytest.mark.skipif(not HAVE_CRYPTO, reason="requires pycryptodome")
 def test_sam_plugin_rev2(target_win_users, hive_hklm):
-    sam_key = VirtualKey(hive_hklm, sam_key_name)
+    sam_key = VirtualKey(hive_hklm, SAM_KEY_PATH)
     sam_key.add_value(
         "F",
         bytes.fromhex(
@@ -301,10 +303,10 @@ def test_sam_plugin_rev2(target_win_users, hive_hklm):
     users_key.add_subkey("000003E8", e8_key)
 
     sam_key.add_subkey("Users", users_key)
-    hive_hklm.map_key(sam_key_name, sam_key)
+    hive_hklm.map_key(SAM_KEY_PATH, sam_key)
 
     # Create SYSTEM keys
-    system_key = VirtualKey(hive_hklm, system_key_name)
+    system_key = VirtualKey(hive_hklm, SYSTEM_KEY_PATH)
     data_key = VirtualKey(hive_hklm, "Data", class_name="36f1befb")
     system_key.add_subkey("Data", data_key)
     gbg_key = VirtualKey(hive_hklm, "GBG", class_name="bae89edb")
@@ -313,7 +315,7 @@ def test_sam_plugin_rev2(target_win_users, hive_hklm):
     system_key.add_subkey("JD", jd_key)
     skew1_key = VirtualKey(hive_hklm, "Skew1", class_name="e6f92d89")
     system_key.add_subkey("Skew1", skew1_key)
-    hive_hklm.map_key(system_key_name, system_key)
+    hive_hklm.map_key(SYSTEM_KEY_PATH, system_key)
 
     target_win_users.add_plugin(SamPlugin)
     results = list(target_win_users.sam())
```

**Comment by JSCU-CNI** [comment]
That's odd, `allow edits by maintainers` is enabled. I've applied your patch to this branch.

**Comment by Schamper** [comment]
If you fix the unit test this is good to go!

**Comment by JSCU-CNI** [comment]
I think the unit test breaks because it is using an old version of `flow.record`:

```
FAILED tests/test_plugins_os_windows_sam.py::test_sam_plugin_rev1 - ValueError: unconverted data remains: +00:00
FAILED tests/test_plugins_os_windows_sam.py::test_sam_plugin_rev2 - ValueError: unconverted data remains: +00:00
```

This should work per https://github.com/fox-it/flow.record/pull/37

**Comment by Schamper** [comment]
Looks like we need to push a new tag for flow.record first then. I'll get that done tomorrow. I already upped the dependency in setup.py.

**Review by Schamper — APPROVED** [review]

**Pull request merged**